homeassistant/components/integration/sensor.py

```diff
@@ -444,13 +501,58 @@ def _handle_state_change(self, event: Event[EventStateChangedData]) -> None:
 
         elapsed_seconds = Decimal(
             (new_state.last_updated - old_state.last_updated).total_seconds()
+            if self._last_integration_trigger == _IntegrationTrigger.StateChange
+            else (new_state.last_updated - self._last_integration_time).total_seconds()
         )
 
         area = self._method.calculate_area_with_two_states(elapsed_seconds, *states)
 
         self._update_integral(area)
         self.async_write_ha_state()
 
+    def _create_on_max_sub_interval_exceeded_callback(
+        self,
+        source_state: State,
+        source_state_dec: Decimal,
+    ) -> Callable[[datetime], None]:
+        @callback
+        def _integrate_on_max_sub_interval_exceeded_callback(now: datetime) -> None:
+            elapsed_seconds = Decimal(
+                (now - self._last_integration_time).total_seconds()
+            )
+            self._derive_and_set_attributes_from_state(source_state)
+            area = self._method.calculate_area_with_one_state(
+                elapsed_seconds, source_state_dec
+            )
+            self._update_integral(area)
+            self.async_write_ha_state()
+
+            self._last_integration_time = datetime.now(tz=UTC)
+            self._last_integration_trigger = _IntegrationTrigger.TimeElapsed
+
+            self._schedule_max_sub_interval_exceeded_if_state_is_numeric(source_state)
+
+        return _integrate_on_max_sub_interval_exceeded_callback
+
+    def _schedule_max_sub_interval_exceeded_if_state_is_numeric(
+        self, source_state: State | None
+    ) -> None:
```

Comment:
Instead of the `_create_on_max_sub_interval_exceeded_callback`, why not just do like this:

```suggestion
    def _schedule_max_sub_interval_exceeded_if_state_is_numeric(
        self, source_state: State | None
    ) -> None:
        @callback
        def _integrate_on_max_sub_interval_exceeded_callback(now: datetime) -> None:
            elapsed_seconds = Decimal(
                (now - self._last_integration_time).total_seconds()
            )
            self._derive_and_set_attributes_from_state(source_state)
            area = self._method.calculate_area_with_one_state(
                elapsed_seconds, source_state_dec
            )
            self._update_integral(area)
            self.async_write_ha_state()

            self._last_integration_time = datetime.now(tz=UTC)
            self._last_integration_trigger = _IntegrationTrigger.TimeElapsed

            self._schedule_max_sub_interval_exceeded_if_state_is_numeric(source_state)

```

Reply:
Nice suggestion!